QoL: Gas Analyzer better TGUI

larentoun opened the pull request:
## About The Pull Request
~~Adds secondary action to Gas Analyzer, which uses TGUI and ignores to_chat() messages.~~
 Gas Analyzer is fully TGUI ([This Commit](https://github.com/tgstation/tgstation/pull/73848/commits/305238fc2083fadcf90be9a783c2e22748020e4b) makes Gas Analyzer fully TGUI)

Adds more info for Gasmix, making TGUI giving the same amount of info as atmos_scan proc.

Gas Analyzer can now target turfs (by giving both ranged and non-ranged version the same afterattack, range changes with var/range).

Added auto-update toggle on TGUI of Gas Analyzer, which allows to auto-scan player turf or targetted with secondary action turf/object.

Added history to TGUI of Gas Analyzer, which can be cleared manually.

To restore to environment scanning, you need to either scan yourself/your tile or activate it

## Why It's Good For The Game
Better TGUI implementation for active scanning ~~,an alternative to primary action usage of Gas Analyzer.~~ ([This Commit](https://github.com/tgstation/tgstation/pull/73848/commits/305238fc2083fadcf90be9a783c2e22748020e4b) overrides this)
Ability to target turfs is handy when there is machinery in a way.

## Changelog
:cl:
add: Gas Analyzer is fully TGUI
add: Additional info for Gasmix TGUI
qol: Gas Analyzer TGUI can automatically update scan data
qol: Gas Analyzer TGUI provides scan history
qol: Gas Analyzer can now target turfs
/:cl:

## Current implementation:
Auto-updating and targetted scans
https://user-images.githubusercontent.com/31931237/225364024-f09fcd7a-b72d-485f-b940-a6707e736d32.mp4

Restoring to scanning environment
https://user-images.githubusercontent.com/31931237/225364074-c6bb2610-9d34-490a-8378-d5609e1e91c4.mp4

larentoun committed:
tweak: Gas Analyzer is now fully TGUI

fix: atmos_scan printing with analyzer

larentoun committed:
add: Additional info for Gasmix TGUI

larentoun committed:
add: Additional info for Gasmix TGUI

larentoun committed:
add: Additional info for Gasmix TGUI

Mothblocks commented:
> instead opens TGUI window (DOES NOT AUTO-UPDATE)

This is going to feel really inconsistent and clunky, so this is definitely a blocker.

It should auto update, but:

- Clicking a pipe/whatever will open a submenu on the tgui for that pipe
- You can click a button to "freeze" current air output

Also this needs pictures.

Pickle-Coding commented:
Sometimes atmospheric technicians might want to monitor the change of the gas mixture over time. The chat spam helped with that because they would be kept in history.

If it opened a tgui for every use, that would be extremely clunky and terrible. If it refreshed for every use, then history gets lost and it might make it harder to monitor the change in values over time.

I suggest adding some way the analyzer can record and keep the information for later use, otherwise I'm not a fan of this.

LemonInTheDark commented:
> If it opened a tgui for every use, that would be extremely clunky and terrible

I also worry about this.

larentoun commented:
Thank you everyone for your input! I was hoping fellow atmostechs would give some insights.

First of all, I should've specified, that TGUI updates with each use (like it prints now in chat), but it will not auto-update values as AtmoZphere does.

Second of all, since I haven't really had a need in History property, I'm a little stuck how to correctly implement it.
The History tab should update:
a) Automatically with each use
b) Manually with a TGUI button "Save scan"

Third, is there really a need in auto-updating values, as AtmoZphere does every UI-tick?
a) Yes, with an ability to "Freeze" data
b) No
If it will auto-update each UI-tick, and history is auto-updated too (point a), wouldn't it be problematic to search for specific history?

I really do hate chat clutter, but I'm looking for a way to implement History for those who might need it.

Fourth, should there be instead a button to print scan data with a cooldown? It could be griefing-potential to spawn a lot of papers tho...

What do you think is the best options for the three questions here?

Mothblocks commented:
> Third, is there really a need in auto-updating values, as AtmoZphere does every UI-tick?

Yes, because the vast majority of in-world UIs are automatically updating. It is completely unintuitive that this one doesn't without manual intervention.

> If it will auto-update each UI-tick, and history is auto-updated too (point a), wouldn't it be problematic to search for specific history?

Yes, which is why this needs a lot more thought than I think it might be worth. I would want some system for seeing rate of change.

> Fourth, should there be instead a button to print scan data with a cooldown? It could be griefing-potential to spawn a lot of papers tho...

There's no griefing potential because you can limit it with a cooldown or make it feed in paper, but I think that would be a separate PR. I don't know if it's actually useful.

GuillaumePrata commented:
We could meet in the middle and let us turn auto update on/off.

The smooth UI auto updating as default for most use cases and being able to switch it off for when we want a static timeline. It won't have a logged history but for most use cases that is not that big of a deal.

Being able to print a paper with those static readings for when you want a logged history would be amazing, paperwork Atmos Tech hype.

Mothblocks commented:
> We could meet in the middle and let us turn auto update on/off.

![image](https://user-images.githubusercontent.com/35135081/223907071-c7df999f-061b-49c8-9ceb-157b2d0cf957.png)

vincentiusvin commented:
Is the chat spam really that big of a deal? 

The UI wasn't intended to be the main way to use this since it blocks a big part of the screen. I just wanted it to be a way for people to open the gas and reaction reference quickly.

I also dont see why it should be auto updating, since reactions happen for a limited amount of time, blink and you'll miss some. 

It being toggleable would solve that guess, but that's annoying to implement, youll need to hold a softref to the gasmix or machinery. Yucky.

It working on clicked tiles is cool tho :+1:

Mothblocks commented:
There is definitely a way to do this right but I'd rather keep the current behavior than do anything suboptimal in comparison

vincentiusvin commented:
> The chat spam is clunky UX and something I would definitely want to avoid

Is it really clunky though? You would need to wait for the window to show up and then resize it everytime you reopen the window. Thats way worse UX.

larentoun commented:
While the discussions happen, I think I will separate a few things from this PR to a new one, for now. Compared to the master's version:
1. Non-range Gas Analyzer targetting adjacent tiles
2. Gas Analyzer TGUI auto-updates with an ability to freeze auto-updates
3. Gasmix update, to allow AtmoZphere and Gas Analyzer TGUI to show the same info as regular Gas Analyzer.

I believe everyone agree on these three being beneficial.

larentoun commented:
Oh, also. I originally planned to disable to_chat() messages, if TGUI is open, but couldn't find a way to check if TGUI is open.

What if there were two ways to use Gas Analyzer?
1. Regular usage, as in master
2. Secondary action (RMB) to open TGUI and NOT print to_chat(), when used with RMB to scan tiles/objects

It could be the best of two worlds: people can still keep the history with LMB/Z action, and just check surrounding area without a need of history (chat spam) with RMB?

larentoun committed:
feat: Gas Analyzer TGUI auto-updates

fix: Gas Analyzer auto-update button only on Tool

ref: removing unused var

fix: grammar fix

larentoun committed:
feat: Gas Analyzer TGUI auto-updates

fix: Gas Analyzer auto-update button only on Tool

ref: removing unused var

fix: grammar fix

larentoun commented:
Gas Analyzer - Targetting tiles!
https://user-images.githubusercontent.com/31931237/224092021-10c4a187-850f-48c6-b4e9-7b9935dc3a06.mp4

Gas Analyzer TGUI - Auto-update (optional)
https://user-images.githubusercontent.com/31931237/224092148-34bd281a-414e-441e-8b9e-dd29e7a23dd8.mp4

@Mothblocks , should I make a different PR for the features above?

Mothblocks commented:
> Is it really clunky though? You would need to wait for the window to show up and then resize it everytime you reopen the window. Thats way worse UX.

It is clunky and you can admit that without saying you would like a tgui, our longer term plan for this was to embed it into the chat. stylemistake had a visual prototype for it that looked really nice.

> I've never had any issues with the chat thing or heard anyone complain about them. Why change what works so well?

I have and I don't think it works well

Mothblocks commented:
I'll look at the others but this PR can't go through if it doesn't auto update by default IMO

Colovorat commented:
having tgui as additional feature would be decent i think, it would give time to playtest this, work on improving it, and the removing to_chat if tgui becomes arguably better with time. or well the opposite
but i ain't coder, just the player who haven't played for a year

Mothblocks commented:
I definitely do not want both to be supported, we need to stick with one

Colovorat commented:
Okay, what if TGUI with 2 modes you can switch between, one is auto updating, and second is updating on click. altough as said above, there is concern of reopenning, resizing and etc

Mothblocks commented:
I don't have a chance to look at the videos right now but my idea is a lock button in the header of the output display to freeze

LemonInTheDark commented:
I need to play around with this. I'm not sure how comfortable I am with a popup anytime you want to analyze something.
It's quite common to just sorta poke at something while you're working, and needing to close/move a window sounds really annoying.

larentoun committed:
Merge branch 'gas_analyzer_better_tgui' of https://github.com/larentoun/tgstation into gas_analyzer_better_tgui

larentoun committed:
feat: Gas Analyzer history

larentoun committed:
feat: Gas Analyzer history clear

larentoun commented:
Changed PR description, added new video example, added a TODO list.

Need to fix UX with history. Someone suggested something around this one
![image](https://user-images.githubusercontent.com/31931237/225096869-27368212-d9d6-464c-8b44-476d66240c8d.png)

Mothblocks commented:
That huge bar at the top should just be a lock/unlock icon

Also that history thing on the right looks fine, make sure to use `toLocaleString()` though so it gets commas

larentoun commented:
> 1. Where should I place the button?
Something like this?
![image](https://user-images.githubusercontent.com/31931237/225328451-0d0c65a4-196c-4a73-b62e-7d6fe4bab5c2.png)
![image](https://user-images.githubusercontent.com/31931237/225328484-9496f12a-fad8-4d46-b93a-b0f63730adc5.png)

larentoun committed:
qol: Better UI/UX for Gas Analyzer TGUI

larentoun committed:
fix: NtosGasAnalyzer up to date

fix: NtosGasAnalyzer layout

larentoun committed:
ref: GasAnalyzerHistory is it's own entity

larentoun committed:
fix: Double-scans fixed, added better "restore to environment scan" description and func

larentoun committed:
feat: Gas Analyzer full TGUI

larentoun commented:
Updated videos on current implementation.
Added a commit which makes it fully TGUI. For test-merges, probably can do without it?

Mothblocks commented:
> Shouldn't there be dots, since it's a floating number? Or do you mean by thousand magnitudes?

As in `1013.6` should show as `1,013.6`, `toLocaleString()` will do that

larentoun commented:
> > Shouldn't there be dots, since it's a floating number? Or do you mean by thousand magnitudes?
> 
> As in `1013.6` should show as `1,013.6`, `toLocaleString()` will do that

![image](https://user-images.githubusercontent.com/31931237/225497811-b4c976f3-599d-43d8-ba79-a83f7bc4fde9.png)
toLocaleString() makes it like this
```(historyViewMode === 'mol'
                  ? allGasmixes[0].total_moles
                  : allGasmixes[0].pressure
                ).toLocaleString()```

Mothblocks commented:
`toFixed` turns the number into a string, `toLocaleString` isn't going to do anything here

The trick is instead `number.toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })`

Mothblocks commented:
Also what country are you from/what locale are you set to? `toLocaleString` chooses based on your locale--those are going to show as commas for everyone else

larentoun committed:
tweak: Gas Analyzer values are now use Locale

larentoun commented:
> Also what country are you from/what locale are you set to? `toLocaleString` chooses based on your locale--those are going to show as commas for everyone else

I never actually knew that Windows had many options considering Locale. Changed values representation to be the same as in English: 
![image](https://user-images.githubusercontent.com/31931237/225638217-da7df4ec-19bf-4439-aa24-c0a45bb9c714.png)

larentoun committed:
ref: deleted unused variable

Mothblocks commented:
Looks better now. Does the scan history automatically make a scrolling frame?

larentoun committed:
ui: Gas Analyzer auto-update is shown by unlock icon

Mothblocks commented:
That video shows the problem--the history shouldn't extend the window size itself, it should contain the history within itself. AKA you scroll through the history, not the window. You can set its container to `overflow-y: scroll` or something.

larentoun commented:
> 

Then I should contain both Gas Content and Gas History to two separate scrollers?

larentoun committed:
ui: gas analyzer scroll bar is for history only

Mothblocks commented:
You got it :+1:

Mothblocks commented:
Ideally the height of the history wouldn't change with the window contents, though, but you got it

larentoun commented:
So, I just wait for the merge? Or should I squash things together into more cohesive commits?

Mothblocks commented:
Don't squash once a PR is up, i'll take a closer look later

LemonInTheDark commented:
> > If it opened a tgui for every use, that would be extremely clunky and terrible
> 
> I also worry about this. Need to see it in use

Tested this out, I'm not sure how I feel. Having the window pop up in your face when you're trying to work on something feels wrong, and gets in the way while looking at pipes.
I often use the analyzer to sort of probe different pipes while doing a reaction, and making it a popup in the screen makes that worse.

I'm not sure how I feel about auto-update, least for pipes and such. Moving away from the pipe and having the thing switch to auto-scan the air around you is real bad, and just in general loosing the control over when the display updates by default makes reading it a bit of a pain. I get the urge to show change over time, to some extent that's why spam clicking is currently a thing, taking samples to get a feel for how it's changing, but that's both more controlled and faster to refresh.

I don't know how relevant the history bit is. It may just be a matter of how I use it but I've never really needed to check it, outside of when I've been debugging atmos issues.

The loss of "good/bad" colors is a major one. not much more to say past that but like, it is useful to know what should not be there at a glance. 

Had a few issues with the history scrollbar autoscrolling my window down to its header. not sure what's going on there.
Also the combining of this and the scan menus leads to some really janky like, tall content which I don't think I like.

https://user-images.githubusercontent.com/58055496/226149552-0fc44847-885e-4940-b689-2ce651afa2f9.mp4

I'm not really convinced this is better then print to chat honestly. 
I'm being a bit negative, I am a boomer after all. I do wish I liked this better. The current to chat isn't great. it's serviceable but the info isn't well contextualized, and we've ended up bloating it a lot which makes reading it hard. I don't really like that, but I don't think this really helps it, and I really am not convinced that just moving the info to tgui makes anything better.

Mothblocks commented:
@LemonInTheDark What if it put the UI in the bottom right corner and not center of the screen? I believe that's possible and we could have it be a part of the /datum/tgui constructor

larentoun commented:
> Tested this out, I'm not sure how I feel. Having the window pop up in your face when you're trying to work on something feels wrong, and gets in the way while looking at pipes. I often use the analyzer to sort of probe different pipes while doing a reaction, and making it a popup in the screen makes that worse.

Ported TGUI on Para saves window location, not sure how to do that on TG.

> I'm not sure how I feel about auto-update, least for pipes and such. Moving away from the pipe and having the thing switch to auto-scan the air around you is real bad, and just in general loosing the control over when the display updates by default makes reading it a bit of a pain. I get the urge to show change over time, to some extent that's why spam clicking is currently a thing, taking samples to get a feel for how it's changing, but that's both more controlled and faster to refresh.

I was not sure how to correctly implement "you are too far", but, probably, could just write it as "you are too far, press this button to scan around"

> I don't know how relevant the history bit is. It may just be a matter of how I use it but I've never really needed to check it, outside of when I've been debugging atmos issues.

I rarely use history myself, but people really liked to go back and look at values

> The loss of "good/bad" colors is a major one. not much more to say past that but like, it is useful to know what should not be there at a glance.

I don't think there ever was one outside of air alarm?

> Had a few issues with the history scrollbar autoscrolling my window down to its header. not sure what's going on there. Also the combining of this and the scan menus leads to some really janky like, tall content which I don't think I like.

I'm not sure what are you talking about either?...

LemonInTheDark commented:
> @LemonInTheDark What if it put the UI in the bottom right corner and not center of the screen? I believe that's possible and we could have it be a part of the /datum/tgui constructor

That would be better (I don't totally trust /datum/tgui cause of window reuse but I could be wrong about how that's handled).
It'd block chat, but people aren't always looking at that when air scanning so it's not so bad.
I'm still a bit rough on why a tgui implementation is better then a chat one to be honest.


> > I'm not sure how I feel about auto-update, least for pipes and such. Moving away from the pipe and having the thing switch to auto-scan the air around you is real bad, and just in general loosing the control over when the display updates by default makes reading it a bit of a pain. I get the urge to show change over time, to some extent that's why spam clicking is currently a thing, taking samples to get a feel for how it's changing, but that's both more controlled and faster to refresh.
> 
> I was not sure how to correctly implement "you are too far", but, probably, could just write it as "you are too far, press this button to scan around"

Maybe. Having it switch autoupdate sources is just.. bad. I don't think you need to prompt the user for scans, just let them use the thing and start updating on that new source. Wouldn't worry bout it too much.

> > I don't know how relevant the history bit is. It may just be a matter of how I use it but I've never really needed to check it, outside of when I've been debugging atmos issues.
> 
> I rarely use history myself, but people really liked to go back and look at values

True, mood.

> > The loss of "good/bad" colors is a major one. not much more to say past that but like, it is useful to know what should not be there at a glance.
> 
> I don't think there ever was one outside of air alarm?

yeah you're right. I just don't play the game and have a fucked memory. scratch this. (It'd be cool tho)

> > Had a few issues with the history scrollbar autoscrolling my window down to its header. not sure what's going on there. Also the combining of this and the scan menus leads to some really janky like, tall content which I don't think I like.
> 
> I'm not sure what are you talking about either?...

When I mouse around on the ui while it auto updates, it will occasionally force my scroll height down to match the top of the history scroll bar. It's hard to see but it feels like shit. That is what's going on in the second half of that video, I never actually scroll down.

larentoun committed:
Merge branch 'master' into gas_analyzer_full_tgui

Mothblocks commented:
It's up to Lemon if he wants this

jlsnow301 commented on tgui/packages/tgui/interfaces/GasAnalyzer.tsx:
```diff
@@ -1,22 +1,47 @@
+  autoUpdating: boolean;
+};
+
+export type GasAnalyzerHistoryData = {
```
you shouldn't need to export this

larentoun replied:
What about `GasAnalyzerData`? If it's not used in another file, it shouldn't have export, right?

jlsnow301 commented on tgui/packages/tgui/interfaces/NtosGasAnalyzer.tsx:
```diff
@@ -1,38 +1,23 @@
 export const NtosGasAnalyzer = (props, context) => {
-  const { act, data } = useBackend<NtosGasAnalyzerData>(context);
-  const { atmozphereMode, clickAtmozphereCompatible } = data;
+  const { act, data } = useBackend<GasAnalyzerData>(context);
```
this doesn't look like it needs to be here

larentoun committed:
ref: Removed unnecessary exports/imports

optimumtact commented:
I think it's clear that the maintainers do not have interest, sorry that this sat for so long

mothblocks and lemon need to do a much better job communicating.

larentoun commented:
As a last resort, I would like to share a picture of a poll result, @Mothblocks @LemonInTheDark 
The same TGUI was ported to ParaCode, and it was met overwhelmingly positively, as seen on the screenshot.

![image](https://user-images.githubusercontent.com/31931237/236618803-293f5066-8451-45f2-a535-539779dda4fc.png)
The translation is:
"Now, a few weeks ago gas anazyler got themselves TGUI, and we would like to get comments about how comfortable it is and what can we do to make it better"
"Green mark - mostly like it, TGUI is superior"
"Red cross - mostly dislike it, text is superior"

People, who voted against it, didn't give comment why.

jlsnow301 commented:
I think this just fell into obscurity as there were no hard votes against it. For your sake I'm going to reopen this and request it to be TM'd. Maybe with some actual usage players will come around.

For the record, I'm indifferent, I don't play atmos but I do think to-chat UIs are on their way out.

jlsnow301 committed:
Merge branch 'master' into gas_analyzer_full_tgui

optimumtact commented:
@jlsnow301 do not reopen this unless you intend to merge it.

jlsnow301 commented:
I'm okay with this if the test merge goes well. I will also close it if it doesn't

jlsnow301 reviewed (changes requested):
My initial concerns with this after having tried it in test is that it isn't well designed - 
1. I like the added information for recipes and reactions, but this is cobbled into the UI and it results in a ton of scrolling. Feels clunky.
2. The sections are blank when nothing is selected where I think they should either not display or display visual information like a noticebox that displays "nothing selected"
3. The analyzer works for ghosts which makes no sense how or why they would need this. And it auto updates as well :|
4. Visual bugs - lack of using full and or grow props causes the readings history to scale height at random while selecting gases/reactions. because of this, scrolling history doesn't work as expected either. you should never have to scroll the page to see the bottom of a scroll bar.
5.  Not your fault, but the readouts are so tightly stacked that it's almost a solid button.
6. You're missing some opportunities here to really make it pop out by using gas colors, which already exist in constants.ts. getGasColor should do this
7. On the bright side, mostly positive-leaning-null impressions on test merge. People don't hate it. Take this as an opportunity to make something useful that people didn't realize they needed until they have it.


I think if you're serious about this UI it needs to be a little bit better thought out. Separate tabs for the info sections would be nice here. Consider that very many users here prefer the old UIs with minimal visual elements and minimal change (myself not included fyi). 

That in mind, this should keep very obvious analyzing functionality intact on the front page and the tertiary helpful info elsewhere.

jlsnow301 commented on tgui/packages/tgui/interfaces/GasAnalyzer.tsx:
```diff
@@ -29,11 +54,85 @@ export const GasAnalyzerContent = (props, context) => {
+        />
+      }>
+      <LabeledList.Item label="Mode">
+        <Flex inline width="50%">
```
use stack, and please do not set width like this, use grow and fill props. [component reference](https://github.com/tgstation/tgstation/blob/master/tgui/docs/component-reference.md)

jlsnow301 commented on tgui/packages/tgui/interfaces/GasAnalyzer.tsx:
```diff
@@ -29,11 +54,85 @@ export const GasAnalyzerContent = (props, context) => {
 
+export const GasAnalyzerHistory = (props, context) => {
+  const { act, data } = useBackend<GasAnalyzerHistoryData>(context);
+  const { historyGasmixes, historyViewMode, historyIndex } = data;
```
```suggestion
  const { historyGasmixes = [], historyViewMode, historyIndex } = data;
```

jlsnow301 commented on tgui/packages/tgui/interfaces/GasAnalyzer.tsx:
```diff
@@ -29,11 +54,85 @@ export const GasAnalyzerContent = (props, context) => {
+          </Flex.Item>
+          <Flex.Item>
+            <Button
+              content={'mol'}
```
do not use content, use put the text as a child

larentoun replied:
What do you mean exactly? I'm kind of new to TGUI

jlsnow301 commented on tgui/packages/tgui/interfaces/GasAnalyzer.tsx:
```diff
@@ -29,11 +54,85 @@ export const GasAnalyzerContent = (props, context) => {
+            <GasAnalyzerContent />
+          </Stack.Item>
+          {/* Right Column */}
+          <Stack.Item width={'160px'}>
```
no width please, grow

jlsnow301 commented on tgui/packages/tgui/interfaces/GasAnalyzer.tsx:
```diff
@@ -29,11 +54,85 @@ export const GasAnalyzerContent = (props, context) => {
-        <GasAnalyzerContent />
+    <Window width={500} height={500}>
+      <Window.Content>
+        <Stack>
```
stack fill

jlsnow301 commented on tgui/packages/tgui/interfaces/GasAnalyzer.tsx:
```diff
@@ -1,22 +1,47 @@
 export const GasAnalyzerContent = (props, context) => {
   const { act, data } = useBackend<GasAnalyzerData>(context);
-  const { gasmixes } = data;
+  const { gasmixes, autoUpdating } = data;
```
the backend does not always send data, so what happens when it tries to use .map on undefined?
```suggestion
  const { gasmixes = [], autoUpdating } = data;
```

jlsnow301 commented on tgui/packages/tgui/interfaces/GasAnalyzer.tsx:
```diff
@@ -1,22 +1,47 @@
+
+type GasAnalyzerData = {
   gasmixes: Gasmix[];
+  autoUpdating: boolean;
```
BooleanLike

LemonInTheDark commented:
@jlsnow301 did you want this? if you disagree with me strongly and think it's better then you can send it. I might feel differently but I hardly play and trust your judgement over mine.
You mention a test merge, does that need done still?

jlsnow301 commented:
> did you want this?

Not in its current state, I had a lot of gripes with it that went unaddressed. The author insisted it would have a positive reception so I gave it a shot with the TM to see what the players think, but that came and went. Appreciate it though. 

If this was reopened later with a decent UI I'd be okay with it.
